cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/handler/RuntimeHandler.java

```diff
@@ -0,0 +1,136 @@
+import co.cask.http.ChunkResponder;
+import co.cask.http.HttpResponder;
+import com.google.common.io.Closeables;
+import com.google.gson.Gson;
+import com.google.gson.reflect.TypeToken;
+import com.google.gson.stream.JsonWriter;
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.ByteBufOutputStream;
+import io.netty.buffer.Unpooled;
+import io.netty.handler.codec.http.DefaultHttpHeaders;
+import io.netty.handler.codec.http.FullHttpRequest;
+import io.netty.handler.codec.http.HttpHeaderNames;
+import io.netty.handler.codec.http.HttpResponseStatus;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.io.OutputStreamWriter;
+import java.lang.reflect.Type;
+import java.nio.charset.StandardCharsets;
+import java.util.Map;
+import javax.annotation.Nullable;
+import javax.ws.rs.POST;
+import javax.ws.rs.Path;
+
+/**
+ * {@link co.cask.http.HttpHandler} for exposing metadata of a runtime.
+ */
+@Path("/v1/runtime")
+public class RuntimeHandler extends AbstractHttpHandler {
+  private static final Logger LOG = LoggerFactory.getLogger(RuntimeHandler.class);
+  private static final Gson GSON = new Gson();
+  private static final Type MAP_STRING_CONSUME_REQUEST_TYPE = new TypeToken<Map<String,
+    MonitorConsumeRequest>>() { }.getType();
+  private static final int CHUNK_SIZE = 8192;
+
+  private final CConfiguration cConf;
+  private final MessageFetcher messageFetcher;
+  private final Runnable shutdownRunnable;
+
+  public RuntimeHandler(CConfiguration cConf, MessageFetcher messageFetcher, Runnable shutdownRunnable) {
+    this.cConf = cConf;
+    this.messageFetcher = messageFetcher;
+    this.shutdownRunnable = shutdownRunnable;
+  }
+
+  /**
+   * Gets list of topics along with offsets and limit as request and returns list of messages
+   */
+  @POST
+  @Path("/metadata")
+  public void metadata(FullHttpRequest request, HttpResponder responder) throws Exception {
+    String requestBody = request.content().toString(StandardCharsets.UTF_8);
+
+    Map<String, MonitorConsumeRequest> consumeRequests = GSON.fromJson(requestBody, MAP_STRING_CONSUME_REQUEST_TYPE);
+
+    ChunkResponder chunkResponder = responder.sendChunkStart(
+      HttpResponseStatus.OK, new DefaultHttpHeaders().set(HttpHeaderNames.CONTENT_TYPE,
+                                                          "application/json; charset=utf-8"));
+    ByteBuf buffer = Unpooled.buffer();
+
+    try (JsonWriter jsonWriter = new JsonWriter(new OutputStreamWriter(new ByteBufOutputStream(buffer),
+                                                                       StandardCharsets.UTF_8))) {
+      jsonWriter.beginObject();
+
+      for (Map.Entry<String, MonitorConsumeRequest> entry : consumeRequests.entrySet()) {
+        jsonWriter.name(entry.getKey());
+        jsonWriter.beginArray();
+
+        try {
+          writeMessages(jsonWriter, buffer, chunkResponder, cConf.get(entry.getKey()), entry.getValue().getLimit(),
```

Comment:
You have to make sure writeMessages doesn't throw, right? Otherwise there could be unsend data in the `buffer`.